### DISCUSSION TITLE
What is the difference between `DocumentNode` vs `Schema` vs `TypeDefs` ?

### DISCUSSION THREAD

**Discussion by hazem-alabiad** [apollographql/apollo-server#5713] [Q&A] [answered]
What is the difference between `DocumentNode` vs `Schema` vs `TypeDefs` ?

**Comment by smolinari** [comment]
A `DocumentNode` or query is the return GraphQL document of the query string (schema/ GraphQL SDL) you use as a client query, which is usually wrapped by `gql`. It's the result of the string literal. It's similar to `typeDefs` on the server, as the queries match closely the typeDefs on the server, in order to get a response from the GraphQL endpoint. 

A `Schema` is what is used to create GraphQL type definitions and client queries. They are written in GraphQL's DSL. 
https://www.apollographql.com/docs/apollo-server/schema/schema/

`TypeDefs` are the definitions of schema, after being ran through the `gql` method or the strings themselves, and used in the Apollo Server to generate the endpoint type definitions (what you also see with introspection in the Playground within Queries and Mutations). In addition to the `typeDefs` you also have `resolvers` (they make up the Queries and Mutations) which use the type definitions to form the response objects and to check for the proper arg or input object types.  

Hope that helped and didn't confuse you even more. 😁 

Scott

**Reply by glasser** [reply]
@smolinari I appreciate your attempt to help here, but this isn't quite correct.  `DocumentNode` is not a string.  `Schema` in this context might mean `GraphQLSchema` which is not DSL.  `TypeDefs` can be ran through `gql` or can still be strings. My response below is a bit more precise.

**Reply by smolinari** [reply]
@glasser - I've updated my reply to me more concise. I've also replied assuming hazemalabiad is a beginner to GraphQL/ Apollo. I also knew how these terms confused me, when I began, and as you can see, I still have a hard time with it. 

Scott

**Comment by glasser** [comment, marked as answer]
There are three types that can represent a GraphQL schema:

- A `string` written in the GraphQL language, often called SDL, like `type Query { field: ID! }`.
- An "abstract syntax tree" representing how a GraphQL language string is parsed. The root of this tree is called `DocumentNode`, because the top level of GraphQL syntax is called `Document`. You can get this from the `graphql-js` `parse` function or the `graphql-tag`/`apollo-server` `gql` function (which is a memoizing version of `parse`).
- A `GraphQLSchema` object, which represents the schema more directly than as a parsed syntax tree. For example, this object contains a map from type name to objects representing that type (but doesn't have the concept of "ordering" between the types), whereas the `DocumentNode` keeps track of the order of declarations but doesn't have an easy way to jump directly from a name to the declaration.  (Also note that `GraphQLSchema` does not contain all information from the AST; for example, it does not track arbitrary directive applications other than via the `astNode` field.) More notably, the `GraphQLSchema` object can have resolver functions attached to it, and the AST and string cannot.

(Also note that GraphQL language strings and `DocumentNode`s can represent operations (queries, mutations, etc) too, whereas `GraphQLSchema` objects are always schemas.)

`typeDefs` is a name given to arguments/options in various places in Apollo Server, not a type like the ones above.  It is an option that takes either a `string` or a `DocumentNode` (or an array of one of those types).
